tools/toolchain/clang_lto_pgo.sh

```diff
@@ -0,0 +1,76 @@
+#!/bin/bash -uex
+
+if [ -d "/optimized_clang" ];
+then
+	DIR="/optimized_clang"
+else
+	DIR="${PWD}/optimized_clang"
+	mkdir -p "${DIR}"
+fi
+cd "${DIR}"
+
+# Download the scylla codebase for training.
+# We assume that Scylla's dependencies are already installed.
+git clone https://github.com/scylladb/scylla --depth=1
+git -C scylla submodule update --init --depth=1
```

Comment:
Why, though? Downloading makes the script more self-contained, and you don't gain anything by using the local sources, other than saving a few MiB of internet traffic.

And how would even you do it? `ADD`ing the local tree to the Docker image is going to leave a garbage layer in overlayfs, and there is no way to interact with the local fs other than ADD. With a script all downloading, copying and removing happens in one build step, so it won't leave a garbage layer.

Reply:
> And how would even you do it? ADDing the local tree to the Docker image is going to leave a garbage layer in overlayfs, and there is no way to interact with the local fs other than ADD. With a script, all downloading, copying and removing happens in one build step, so it won't leave a garbage layer.

You can do this by building clang in a separate image (e.g. using the multi-stage build feature) and only copying the resulting clang binary to the main image (which is a good idea anyway), but I still don't see the point.

Reply:
The current setup is not reproducible. Say you have a year-old branch, you update it to get a vulernerability patched, but you can't build the compiler because scylladb.git master doesn't build with this older compiler.

regarding Dockerfile `ADD`, if it's bad, let's use something else.

Reply:
btw, there's --squash, (see tools/toolchain/prepare), so no garbage layers.

Reply:
> The current setup is not reproducible. Say you have a year-old branch, you update it to get a vulernerability patched, but you can't build the compiler because scylladb.git master doesn't build with this older compiler.

Good point. But that's fixable simply by downloading some fixed version, rather than master. It shouldn't matter what exactly version on Scylla you are training on. You can update the training workload once every 3 years, or when the build breaks due to a new compiler version.